core/config/v2/validate.go

```diff
@@ -16,24 +16,31 @@ type Validated interface {
 	// ValidateConfig returns nil if the config is valid, otherwise an error describing why it is invalid.
 	//
 	// For implementations:
-	//  - A nil receiver should return nil, freeing the caller to decide whether each case is required.
 	//  - Use package multierr to accumulate all errors, rather than returning the first encountered.
+	//  - If an anonymous field also implements ValidateConfig(), it must be called explicitly!
 	ValidateConfig() error
 }
 
 // Validate returns any errors from calling Validated.ValidateConfig on cfg and any nested types that implement Validated.
 func Validate(cfg interface{}) (err error) {
-	_, err = utils.MultiErrorList(validate(cfg))
+	_, err = utils.MultiErrorList(validate(reflect.ValueOf(cfg), true))
 	return
 }
 
-func validate(s interface{}) (err error) {
-	if vc, ok := s.(Validated); ok {
-		err = multierr.Append(err, vc.ValidateConfig())
+func validate(v reflect.Value, checkInterface bool) (err error) {
+	if checkInterface {
+		i := v.Interface()
+		if vc, ok := i.(Validated); ok {
+			err = multierr.Append(err, vc.ValidateConfig())
+		} else if v.CanAddr() {
+			i = v.Addr().Interface()
+			if vc, ok := i.(Validated); ok {
+				err = multierr.Append(err, vc.ValidateConfig())
+			}
+		}
```

Comment:
This is the only meaningful change. Now we check if a pointer to the value implements validation, and we also skip either check when the value is an anonymous struct field, since the method could be inherited from the embedding and we need to avoid a double call.